pkg/codec/codec.go

```diff
@@ -8,5 +8,35 @@ import (
 	"github.com/pion/mediadevices/pkg/prop"
 )
 
-type VideoEncoderBuilder func(r video.Reader, p prop.Media) (io.ReadCloser, error)
-type AudioEncoderBuilder func(r audio.Reader, p prop.Media) (io.ReadCloser, error)
+// AudioEncoderBuilder is the interface that wraps basic operations that are
+// necessary to build the audio encoder.
+//
+// This interface is for codec implementors to provide codec specific params,
+// but still giving generality for the users.
+type AudioEncoderBuilder interface {
+	// Name represents the codec name
+	Name() string
+	// BuildAudioEncoder builds audio encoder by given media params and audio input
+	BuildAudioEncoder(r audio.Reader, p prop.Media) (io.ReadCloser, error)
+}
+
+// VideoEncoderBuilder is the interface that wraps basic operations that are
+// necessary to build the video encoder.
+//
+// This interface is for codec implementors to provide codec specific params,
+// but still giving generality for the users.
+type VideoEncoderBuilder interface {
+	// Name represents the codec name
+	Name() string
+	// BuildVideoEncoder builds video encoder by given media params and video input
+	BuildVideoEncoder(r video.Reader, p prop.Media) (io.ReadCloser, error)
+}
+
+// BaseParams represents an codec's encoding properties
+type BaseParams struct {
+	// Target bitrate in bps.
+	BitRate int
```

Comment:
One concern is that it is not enforced to embed `codec.BaseParams` to `EncoderBuilder`.
It would be nice to have a unified method to set bit-rate and force key-framing. (It will be required to dynamically set bit-rate to realize WebRTC's keyframe request and bandwidth control.)

For example,
```go
type StreamReader interface {
	io.ReadCloser
	SetBitRate(int) error
	ForceKeyFrame() error
}

BuildVideoEncoder(r video.Reader, p prop.Media) (StreamReader, error)
```

It may be better to be done in different PR.

Reply:
Yep, this is definitely the vision that I had too. And, yes, I think we should separate it into a different PR.